---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```
# QCnormSE

<!-- badges: start -->
<!-- badges: end -->

The QCnormSE package provides implementations of several quality control 
procedures and processing steps for normalized gene expression data sets from 
microarray and RNA-seq experiments in a common framework. 

## Installation

First install the dependencies from CRAN:

```{r, eval = FALSE}
## install CRAN dependencies
cran.dep = c("aplpack",
             "caret",
             "circlize",
             "GGally",
             "ggplot2",
             "ggplotify",
             "ggpubr",
             "Hmisc",
             "tidyr")

cran.dep.to.install = setdiff(cran.dep,
                     installed.packages()[, "Package"])

if (length(cran.dep.to.install) > 0) {
    install.packages(cran.dep.to.install)
}
```

Next install the dependencies from Bioconductor:

```{r, eval = FALSE}
## install Bioconductor dependencies
bioc.dep = c("affyio",
             "BiocGenerics",
             "ComplexHeatmap",
             "edgeR",
             "doppelgangR",
             "GEOquery",
             "illuminaio",
             "recount",
             "S4Vectors",
             "SummarizedExperiment")

bioc.dep.to.install = setdiff(bioc.dep,
                              installed.packages()[, "Package"])

if (length(bioc.dep.to.install) > 0) {
    if (!requireNamespace("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
    
    BiocManager::install(bioc.dep.to.install)
}

```

Then install the most recent version of QCnormSE from GitHub:

```{r, eval = FALSE}
library(devtools)
install_github("szymczak-lab/QCnormSE")
```

## Usage

A detailed user guide is available as vignette.
